Add GridTileComponent and base theming syntax

Koerty opened the pull request:
This is biggest PR I created yet (since I started to completely disassemble the grid view), so take a big breath before you dive into the code.

**Changelog**
- Add the GridTileComponent which hold the image and its background
- Add base theming syntax for the ImageGrid and GridTile
- Numerous refactoring/cleaning in ImageGridComponent
  - GridDimension is not recalculated dynamically, we don't need that with new grid display algorithm, so we just (re)calculate it when ImageGridComponent got instantiated or after theme changed
  - The ImageGridComponent control the GridTiles only using setImage, setSelected and setVisible, all the tile parameters (max size, image color, ...) are stored inside the GridTileComponent class
  - Moved all the maths for calculating the start position (the number of the tile to display first on top left of the screen) from the updateImages function to a new function called getStartPosition

**Theming syntax**
Full allowed theming syntax after this changes (with default values) : 
```xml
<view name="grid">
	<imagegrid name="gamegrid">
		<pos>0.1 0.1</pos>
		<size>0.80 0.80</size>
		<margin>0.07 0.07</margin>
	</imagegrid>

	<gridtile name="default">
		<size>0.22 0.22</size>
		<padding>16 16</padding>
		<imageColor>0xAAAAAABB</imageColor>
		<backgroundPath>":/frame.png"</backgroundPath>
		<backgroundColor>0xAAAAEEFF</backgroundColor>
	</gridtile>

	<gridtile name="selected">
		<size>0.264 0.264</size>
		<padding>16 16</padding> <!-- not based on screen size unlike other sizes -->
		<imageColor>0xFFFFFFFF</imageColor>
		<backgroundPath>":/frame.png"</backgroundPath>
		<backgroundColor>0xFFFFFFFF</backgroundColor>
	</gridtile>

	<!-- All the metadata stuff from detailed view -->
</view>
```

You can find my latest and final theming syntax proposition, and how the selected and default grid tile parameters interact with each others, here : https://github.com/RetroPie/EmulationStation/issues/206#issuecomment-379541647

jrassa commented:
@Koerty I haven't yet had a chance to test out the code, but wanted to share a couple of quick initial thoughts.

* Should the `maxSize` property on `gridtile` be named `size`.  Presumedly, the grid tile (including background) will always be that size, while `maxSize` really only applies to the image within the tile.
* The proposal in #206 also lists a `padding` property for gridtile that is not present here.  Without this, or a default padding, wouldn't the image in the tile potentially overlap/cover the tile background.

I may have some additional thoughts once I get a chance to test this out,  but overall, it looks good and I am happy with the theme syntax.

Koerty commented:
@jrassa 
> Should the maxSize property on gridtile be named size. Presumedly, the grid tile (including background) will always be that size, while maxSize really only applies to the image within the tile.

The image size is = to the tile size minus the padding, while keeping aspect ratio.
The background is = to the image size plus de padding.

I tried your solution first, setting the background to the size of the tile, but it give way too bad visual so I decided to change the default behavior, but may reintroduce this later on with a new theming element.

![alt text](https://user-images.githubusercontent.com/24305945/38314094-c98fd3c2-3825-11e8-8556-1576e864af6c.png)

>The proposal in #206 also lists a padding property for gridtile that is not present here. Without this, or a default padding, wouldn't the image in the tile potentially overlap/cover the tile background.

The padding isn't here because I had some problems with the NinePatchComponent, which already have natural padding, and can't be changed (I tried to modify the NinePatchComponent to be able to change its padding, but it completly broke). So I decided to delay the introduction of the padding property until I find a proper way to handle the padding of both the NinePatchComponent and the GridTileComponent. This PR is already big enough anyway.

Note the backgroundCenterColor and backgroundEdgeColor aren't here for the exact same reason, it depend on the padding of the NinePatchComponent.

Koerty commented:
Forgot to say in my first message that this still need more testing. But I spent quite some time cleaning and testing the code so you can start reviewing it, it wont change much.

Koerty commented:
There is a strange bug. Right now I resize the images in the update function, and there is one image which is constantly "moving", by that I mean it size is usually 146x97px, but sometimes one or 2 pixels more in size, this come from the line 138 of GridTileComponent.cpp : 
```
mImage->setMaxSize(currentProperties.mMaxSize - Vector2f(32.0f, 32.0f));
```

I don't know if this is the operations between the 2 Vector2f which don't give consistent results, or the rounding which is done later to have the number of pixels, but it is strange.

It only happen on ONE image on the 100 I have.

The resize in the update function isn't necessary actually, as I already do it in the setImage function, line 120. So this is not a critical bug since it is easy to workaround, but I'm curious in why it is happening.

Koerty commented:
Also, the backgroundColor may be uneeded, this is just a way for themes to set both the backgroundEdgeColor and backgroundCenterColor at the same time.

Sorry for spamming, I'm just writing messages when I think about stuff so I don't forget.

jrassa commented:
> The image size is = to the tile size minus the padding, while keeping aspect ratio.
The background is = to the image size plus the padding.

I feel like this may need to be revisited.  I did some testing and to me it looks unbalanced that the grid tiles are not a consistent size.  The margins between the tiles are inconsistent now, and I thought part of the purpose of the background was to mask/hide the inconsistent margins that can occur due to preserving the aspect ratios of the images.  I understand that you did not like the excessive white space on the sides of the boxart in your sample image above, but I think that could have been fixed by changing the tile size.

Including a couple of images to illustrate what I am seeing.
<img width="641" alt="screen1" src="https://user-images.githubusercontent.com/23661980/38533384-0e4b3f40-3c47-11e8-859c-056911d13826.png">

<img width="640" alt="screen2" src="https://user-images.githubusercontent.com/23661980/38534091-39fb911e-3c4a-11e8-9e19-9dd977db0837.png">

Koerty commented:
> I understand that you did not like the excessive white space on the sides of the boxart in your sample image above, but I think that could have been fixed by changing the tile size.

This mean the theme should assume the user will only use the box arts (no screenshot or other stuff like that) to provide a grid with different tile sizes for different systems. Isn't this argument contradictory with everything you said on the grid view subject since the beginning ?

I could provide a new theming element, `fitBackgroundTo=image|tile`, that wouldn't add too much extra code, but that wouldn't change much about our pb.

Screenshots to compare with no theme, material theme, moderntv theme.

![capture d ecran de 2018-04-10 12-47-48](https://user-images.githubusercontent.com/24305945/38553096-7d1d45ea-3cbe-11e8-80eb-5a4c8cac44b5.png)

![capture d ecran de 2018-04-10 12-44-14](https://user-images.githubusercontent.com/24305945/38553100-821a87a6-3cbe-11e8-92e9-b70b31f8dc92.png)

![capture d ecran de 2018-04-10 12-47-08](https://user-images.githubusercontent.com/24305945/38553110-871c4d70-3cbe-11e8-9df6-f8e4541c07fa.png)

![capture d ecran de 2018-04-10 12-46-09](https://user-images.githubusercontent.com/24305945/38553118-89ab513a-3cbe-11e8-8a82-54f749c450de.png)

![capture d ecran de 2018-04-10 12-47-32](https://user-images.githubusercontent.com/24305945/38553141-93618faa-3cbe-11e8-9c59-14764b9a8c37.png)

![capture d ecran de 2018-04-10 12-42-17](https://user-images.githubusercontent.com/24305945/38553146-98ac1688-3cbe-11e8-9791-596f7af3abc9.png)

My feeling is if you hide the background of unselected tiles, it's ok. If you don't, it look kinda garbage.

jrassa commented:
Let me start off by saying that I don't expect this to be solved now.  It is now going to prevent me from merge this when it is ready, but I do think it needs to be addressed before grid view is made available to the masses.

> This mean the theme should assume the user will only use the box arts (no screenshot or other stuff like that) to provide a grid with different tile sizes for different systems. Isn't this argument contradictory with everything you said on the grid view subject since the beginning ?

Thats not what I am trying to say.  The current tile size in your screens matches the aspect ratio of the screen which is much wider than any of the images being used.  Adjusting the tile size to more closely match the widest image used (screenshots) I think would yield a better result.  There would still be excess white space on the sides of the box art images, but not nearly as much.

> My feeling is if you hide the background of unselected tiles, it's ok. If you don't, it look kinda garbage.

I agree.  For me I think it is that visually it doesn't bother me that the images are not the same size because they are natural different and that makes sense.  The background however is visually the same for each tile so the difference in size sticks out more.

I think part of the problem now is the default white background. It makes the excess space really standout.  I expect themes to change the background to something more visually pleasing.

I guess the other issue is that as implemented now, depending on the image type, the background functionally is just a border.

> I could provide a new theming element, fitBackgroundTo=image|tile, that wouldn't add too much extra code, but that wouldn't change much about our pb.

This could be an option to consider.  For now I think we should just get some feedback on the forums.

Koerty commented:
> This could be an option to consider. For now I think we should just get some feedback on the forums.

Yes exactly, I was preparing a new post for the forum which clearly but shortly introduce the problem. I will tag you in when I'm done.

Koerty commented:
@jrassa Small resume of the situation after discussing it on the forum : 

I would like, if possible, to give 3 possibilities to the theme : 
- Fit the image inside the tile + stretch the background to the tile size
- Fit the image inside the tile + stretch the background to the image size
- Resize the image so it completely fill the tile + cut the extra part if the tile is unselected + stretch the background to the image size

![](https://retropie.org.uk/forum/assets/uploads/files/1523396083790-capture-d-%C3%A9cran-de-2018-04-10-23-32-00-resized.png)

The biggest problem is the 3rd solution look really nice, but the selected tile can easily go out of the screen or on top of other elements if it's in the first/last row/column.

I wanted to add some code so the tile only grow in 3 directions when in first/last row/column, 2 directions when in a corner, but this is way harder than expected, I'm only trying to make it work for the 3rd solution, as I removed the code for the 2 firsts, and I'm already struggling.

I fear that all of this combined will create completely unreadable code and impossible to maintain (not even mentioning reintroducing the text).

So we have somehow, to exclude some possibilities.

hex007 commented:
I really like how it looks in the screen shot in the previous image. Its way better to look, very concise, preserves vertical space, is constrained by the images width rather than height.

jrassa commented:
@Koerty Thanks for the summary.  I have been following along on the forums.  I have to say, in your sample images on the forum, the ones with the genesis boxart, the selected image seems to big to me, but since it will be optional I am ok with it.

> I would like, if possible, to give 3 possibilities to the theme :
> 
> Fit the image inside the tile + stretch the background to the tile size
> Fit the image inside the tile + stretch the background to the image size
> Resize the image so it completely fill the tile + cut the extra part if the tile is unselected + stretch the background to the image size

I think we may be able to simplify this to 2 options in the theme.  (Names are just off the top of my head and I am open to changes)

* imageStyle
  * specifies how the image should be rendered within the tile.
  * possible values
    * original (corresponds to setMaxSize method on ImageComponent)
    * cropped (ideally would be implemented as new setCropSize method on ImageComponent)
    * stretched (corresponds to setResize method on ImageComponent)
backgroundStyle
  * specifies how the background should be rendered within the tile
  * possible values
      * fitToImage (set background size to image size + padding)
      * fitToTile (set background size to tile size)
      * fitToFullImage (set background size to large image size for use case 3)

These 2 options should handle the first 2 possibilities you listed and I think that it should handle the 3rd as well.

> I wanted to add some code so the tile only grow in 3 directions when in first/last row/column, 2 directions when in a corner, but this is way harder than expected, I'm only trying to make it work for the 3rd solution, as I removed the code for the 2 firsts, and I'm already struggling.

I'm not sure how you have approached it so far, but I would suggest trying to leverage the origin value.  I believe currently you are setting the origin of the tiles to `0.5 0.5` which would be the center of the tile,

For a 3x3 grid, you probably need the origins to be set to something along the lines of the following.

`0.25 0.25` `0.5 0.25` `0.75 0.25` 
`0.25 0.50` `0.5 0.50` `0.75 0.50`
`0.25 0.75` `0.5 0.75` `0.75 0.75`

I didn't actually test out these values they are just guesses at values that think would work.

jrassa commented:
Regardless, of all of that if you want to hold off that work for a subsequent PR that is fine and I can finish up the review of this as is when you think it is ready.

Also, I can take a look at implementing the setCropSize method for ImageComponent if you want.

Koerty commented:
> I'm not sure how you have approached it so far, but I would suggest trying to leverage the origin value. I believe currently you are setting the origin of the tiles to 0.5 0.5 which would be the center of the tile,

I wasn't satisfied by toying with origins so I tried to do it with position but failed, I will take another look later.

> Regardless, of all of that if you want to hold off that work for a subsequent PR that is fine and I can finish up the review of this as is when you think it is ready.

Yes sure let's move ahead on this PR.

> Also, I can take a look at implementing the setCropSize method for ImageComponent if you want.

I already did that. I added 2 Vector2f called mTopLeftCrop (0, 0 by default) and mBottomRightCrop (1, 1 by default), I added 4 functions setTopCrop setLeftCrop .... which change this vectors.

I multiplied the good vertice values by this vectors in the updateVertices function and it work.

You can check my branch "grid-tile-min-size", please don't mind the disorder, this is far from "PR ready" ;)

Koerty commented:
I did all the possible tests I had in mind, made the requested changes and rebased. For me this is ready to get merged if you don't have any additional change to request.

jrassa commented:
Looks good.  Thanks.
